storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java

```diff
@@ -239,26 +251,45 @@ public IAtomContainer readConnectionTable(IChemObjectBuilder builder) throws CDK
         // read in any SDF fields
         if (lastLine != null && lastLine.startsWith(M_END)) {
             try {
-                MDLV2000Reader.readNonStructuralData(input, state.mol);
+                MDLV2000Reader.readNonStructuralData(input, state.atomContainer);
             } catch (IOException ex) {
                 throw new CDKException("IO Error", ex);
             }
         }
 
-        finalizeMol(state);
+        // carry out final processing steps
+        readData = finalizeMol(state);
 
         return readData;
     }
 
-    private void finalizeMol(ReadState state) {
+    /**
+     * Finalizes the molecule. This includes finalizing dimensions, query features, valence, and stereochemistry.
+     *
+     * @param state the ReadState object containing the molecule and other relevant information
+     * @return the processed Molecule
+     */
+    private IAtomContainer finalizeMol(ReadState state) {
         finalizeDimensions(state);
 
-        IAtomContainer readData = state.mol;
-        boolean isQuery = readData instanceof IQueryAtomContainer;
+        // finalize query features
+        IAtomContainer readAtomContainer = state.atomContainer;
+        // migrate atom container to IQueryAtomContainer implementation if there are any objects with query features
+        if (state.isQuery) {
+            // shallow copy of the original atom container, i.e. same atoms and electron containers as original
+            final IQueryAtomContainer queryAtomContainer = new QueryAtomContainer(readAtomContainer, readAtomContainer.getBuilder());
```

Comment:
This isn't quite right, it will work but I would "deref" the atoms and bonds from the initial since IQueryAtomContainer isn't implemented in terms of that. 

Ideally in future IQueryAtomContainer will go away or at least have better internals. Right now it's not a priority as performance of query manipulation isn't that critical.

Reply:
Not quite sure what you mean here 🤷🏼

Reply:
So the current non-query AtomContainer's wrap every atom and bond in a ref. This is basically a hack to allow Atoms and Bonds to be added to a multiple atom containers yet when you access an atom/bond from a container it knows where it came from. Hence you can ask things like "atom.bonds()" and it works.

The QueryAtomContainer is "dump" and doesn't provide this semantics. 
Option 1) Unbox/deref the atoms/bonds before they go in (quick)
Option 2) Make the QueryAtomContainer smarter (best)

I'll do option 1 by a separate PR.

Reply:
If I was designing the API from scratch I would not allow an atom/bond to be in multiple Molecules at once, it's not as much of a benefit as allowing each atom/bond to be "owned" by a single mol.

Reply:
Fixed it now in #1105